### DISCUSSION TITLE
V14.0.3 Immer wieder mal neg. Werte

### DISCUSSION THREAD

**Discussion by sammykd** [jomjol/AI-on-the-edge-device#2007] [General] [answered]
Hallo, setzte die Version 14.0.3 auf der Wasseruhr ein.
Erhalte immer mal wieder neg. Werte und habe keine Idee mehr woran das liegen kann.
Habe schon auf ana-cont_11.3.1_s2.tflite umgestellt, wo bei es sich dann gebessert hat, aber nicht weg ist, oder muss man damit leben, dass das so ist, habe keine Erfahrungen mit anderen Wasseruhren. bei meiner  GASuhr läuft es reibungslos.
Gibt es noch einen Parameter den ich verbessern kann?
Oder läuft die Uhr ggf. kurz rückwärts, durch den Pufferspeicher des Warmwassersystems, es tritt aber immer nur bei der letzten Stelle bei einer 8 oder 9 auf 

Auszug aus dem Log:
```2023-02-10T11:37:10+0100,main,0508.0112,508.0112,508.0112,0.000140,0.0007,no error,0.0,5.0,0.0,8.0,0.0,1.2,1.4,2.7
2023-02-10T11:42:10+0100,main,0508.0120,508.0120,508.0120,0.000160,0.0008,no error,0.0,5.0,0.0,8.0,0.0,1.4,2.1,0.6
2023-02-10T11:47:10+0100,main,0508.0123,508.0123,508.0123,0.000060,0.0003,no error,0.0,5.0,0.0,8.0,0.0,1.5,2.4,3.8
2023-02-10T11:52:10+0100,main,0508.0123,508.0123,508.0123,0.000000,0.0000,no error,0.0,5.0,0.0,8.0,0.0,1.5,2.4,3.8
2023-02-10T11:57:10+0100,main,0508.0129,508.0129,508.0129,0.000120,0.0006,no error,0.0,5.0,0.0,8.0,0.0,1.5,2.9,9.0
2023-02-10T12:02:10+0100,main,0508.0129,508.0129,508.0129,0.000000,0.0000,no error,0.0,5.0,0.0,8.0,0.0,1.5,2.9,9.0
2023-02-10T12:07:10+0100,main,0508.0128,,508.0129,,0.0000,Neg. Rate - Read: - Raw: 0508.0128 - Pre: 508.0129 ,0.0,5.0,0.0,8.0,0.1,1.5,2.9,9.0

2023-02-10T12:12:10+0100,main,0508.0218,508.0218,508.0218,0.001780,0.0089,no error,0.0,5.0,0.0,8.0,0.1,2.3,2.1,8.8
2023-02-10T12:17:10+0100,main,0508.0242,508.0242,508.0242,0.000480,0.0024,no error,0.0,5.0,0.0,8.0,0.1,2.5,4.1,2.6
2023-02-10T12:22:10+0100,main,0508.0242,508.0242,508.0242,0.000000,0.0000,no error,0.0,5.0,0.0,8.0,0.2,2.5,4.1,2.6`
```
Anbei meine Konfigurationj:
```[MakeImage]
;LogImageLocation = /log/source
WaitBeforeTakingPicture = 5
;LogfileRetentionInDays = 15
;Demo = true
Brightness = 1
Contrast = 2
Saturation = 0
LEDIntensity = 17
ImageQuality = 12
ImageSize = VGA
FixedExposure = false

[Alignment]
InitialRotate = 179
InitialMirror = false
SearchFieldX = 20
SearchFieldY = 20
AlignmentAlgo = default
FlipImageSize = false
/config/ref0.jpg 128 229
/config/ref1.jpg 241 28

[Digits]
Model = /config/dig-cont_0600_s3.tflite
;CNNGoodThreshold = 0.5
;LogImageLocation = /log/digit
;LogfileRetentionInDays = 3
main.dig1 210 107 30 52 false
main.dig2 247 107 30 52 false
main.dig3 284 107 30 52 false
main.dig4 321 107 30 52 0

[Analog]
Model = /config/ana-cont_11.3.1_s2.tflite
;LogImageLocation = /log/analog
;LogfileRetentionInDays = 3
main.ana1 404 181 97 93 false
main.ana2 362 285 94 96 false
main.ana3 257 328 93 96 false
main.ana4 145 282 95 96 false

[PostProcessing]
main.DecimalShift = 0
;main.AnalogDigitalTransitionStart = 9.2
PreValueUse = true
PreValueAgeStartup = 720
main.AllowNegativeRates = false
main.MaxRateValue = 0.5
;main.MaxRateType = AbsoluteChange
;main.ExtendedResolution = false
;main.IgnoreLeadingNaN = true
ErrorMessage = true
CheckDigitIncreaseConsistency = false

[MQTT]
Uri = mqtt://192.168.17x.x:1883  (ausgeXt)
MainTopic = wasserzaehler
ClientID = Wasserzaehler
user = zxy
password = xyz
;SetRetainFlag = true
;HomeassistantDiscovery = true
;MeterType = other

;[InfluxDB]
;Uri = undefined
;Database = 
;Measurement = undefined
;user = undefined
;password = undefined

;[GPIO]
;IO0 = input disabled 10 false false 
;IO1 = input disabled 10 false false 
;IO3 = input disabled 10 false false 
;IO4 = built-in-led disabled 10 false false 
;IO12 = input-pullup disabled 10 false false 
;IO13 = input-pullup disabled 10 false false 
LEDType = WS2812
LEDNumbers = 2
LEDColor = 150 150 150

[AutoTimer]
AutoStart = true
Intervall = 5

[DataLogging]
DataLogActive = true
DataLogRetentionInDays = 3

[Debug]
Logfile = 1
LogfileRetentionInDays = 3

[System]
TimeZone = CET-1CEST,M3.5.0,M10.5.0/3
TimeServer = pool.ntp.org
;AutoAdjustSummertime = false
;Hostname = undefined
;RSSIThreashold = 
SetupMode = false
```
![image](https://user-images.githubusercontent.com/57226083/218084230-8b86ef3c-d798-458a-85fe-b2a7339fb551.png)

**Comment by caco3** [comment]
Ich habe bei mir gute Erfahrungen mit dem `ana-clas100` Model.
Aber auch ich sehe hin und wieder `Neg. Rate`, weil die 4  Stelle  manchmal wieder etwas zurückspringt.
Rückwärts wird sie wohl kaum drehen!

Der Parameter [`ExtendedResolution`](https://jomjol.github.io/AI-on-the-edge-device-docs/Parameters/#PostProcessing-NUMBER.ExtendedResolution) sollte da etwas helfen. (Soviel ich weiss, funktioniert das aber nur mit `*-class*` und `*-const` Modellen!

Siehe auch https://github.com/jomjol/AI-on-the-edge-device/issues/2010

**Reply by sammykd** [reply]
Hallo caco3, 
danke für die Info, habe ich gestern mal aktiviert, bis heute ist alles OK gelaufen. Werde es weiter beobachten., ggf, dann noch mal das andere Model versuchen.

**Reply by Piti61** [reply]
> Rückwärts wird sie wohl kaum drehen!

Meine Wasseruhr läuft tatsächlich gelegentlich auf der vierten Nachkommastelle so 3-6 Zähler rückwärts und anschließend wieder in etwa auf den vorhergehenden Zählerstand vor: selbst beobachtet und mit Video-Kamera dokumentiert.

**Reply by caco3** [reply]
also so 3..6 LDeziliter?
echt komisch!
du könntest diesen Zeiger auch ganz weglassen, da es sowieso nur halbwegs ausagekräftig ist.

**Reply by Piti61** [reply]
> also so 3..6 LDeziliter?

Ja, richtig! In anderen Worten etwa 0,3 bis 0,6 Liter

**Comment by sammykd** [comment, marked as answer]
Hallo caco3,
bis heute 436 Durchläufe ohne Fehler, denke das die Einstellungsergänzung geholfen hat.
Danke  noch mal

**Comment by fm-maniac** [comment]
Hallo,

ich habe hier das gleiche Problem und bekomme es nicht gelöst. Sehr häufig werden auf den analogen Anzeigen zu hohe Werte erkannt, was dazu führt das über lange Zeiträume keine aktuellen Werte mehr kommen. Da brauche ich nun definitiv Hilfe, wie ich das verbessern kann, denn so ist das für mich nicht nutzbar.

Falls weitere Info benötigt wird bitte anfordern!

Grüße!

![2023-02-20 14_23_27-Window](https://user-images.githubusercontent.com/113193480/220122249-76fc8316-9e90-4b22-837b-1a3d907b9119.png)
![2023-02-20 14_25_01-Window](https://user-images.githubusercontent.com/113193480/220122243-64528283-b586-4936-9941-68b5df096800.png)
[data_2023-02-20.csv](https://github.com/jomjol/AI-on-the-edge-device/files/10784941/data_2023-02-20.csv)

```

[MakeImage]
;LogImageLocation = /log/source
WaitBeforeTakingPicture = 5
;LogfileRetentionInDays = 15
;Demo = false
Brightness = -2
Contrast = 0
Saturation = 0
LEDIntensity = 50
ImageQuality = 12
ImageSize = VGA
FixedExposure = false

[Alignment]
InitialRotate = 0
InitialMirror = false
SearchFieldX = 20
SearchFieldY = 20
AlignmentAlgo = default
FlipImageSize = false
/config/ref0.jpg 307 88
/config/ref1.jpg 430 159

[Digits]
Model = /config/dig-cont_0600_s3.tflite
CNNGoodThreshold = 0.5
;LogImageLocation = /log/digit
;LogfileRetentionInDays = 3
main.dig1 247 147 27 46 false
main.dig2 282 147 27 46 false
main.dig3 318 147 27 46 false
main.dig4 354 147 27 46 0
main.dig5 389 147 27 46 0

[Analog]
Model = /config/ana-cont_11.3.1_s2.tflite
;LogImageLocation = /log/analog
;LogfileRetentionInDays = 3
main.ana1 424 228 70 70 false
main.ana2 382 305 70 70 false
main.ana3 308 337 70 70 false
main.ana4 212 301 70 70 false

[PostProcessing]
main.DecimalShift = 0
main.AnalogDigitalTransitionStart = 9.2
PreValueUse = true
PreValueAgeStartup = 720
main.AllowNegativeRates = false
main.MaxRateValue = 0.02
main.MaxRateType = RateChange
main.ExtendedResolution = true
;main.IgnoreLeadingNaN = true
ErrorMessage = true
CheckDigitIncreaseConsistency = false

[MQTT]
Uri = mqtt://192.168.178.40:1883
MainTopic = Wasserzaehler
ClientID = Nutzwasser
user = watermeter
password = ***
SetRetainFlag = false
HomeassistantDiscovery = false
MeterType = water_m3

;[InfluxDB]
;Uri = undefined
;Database = 
;Measurement = undefined
;user = undefined
;password = undefined

;[GPIO]
;IO0 = input disabled 10 false false 
;IO1 = input disabled 10 false false 
;IO3 = input disabled 10 false false 
;IO4 = built-in-led disabled 10 false false 
;IO12 = input-pullup disabled 10 false false 
;IO13 = input-pullup disabled 10 false false 
LEDType = WS2812
LEDNumbers = 2
LEDColor = 150 150 150

[AutoTimer]
AutoStart = true
Intervall = 3

[DataLogging]
DataLogActive = true
DataLogRetentionInDays = 3

[Debug]
Logfile = 1
LogfileRetentionInDays = 3

[System]
TimeZone = CET-1CEST,M3.5.0,M10.5.0/3
;TimeServer = pool.ntp.org
;AutoAdjustSummertime = false
;Hostname = undefined
;RSSIThreashold = 
SetupMode = false
```

**Reply by caco3** [reply]
Hast du es mal mit `class100` Modellen versucht?

**Reply by fm-maniac** [reply]
Ja, ich habe aktuell `class100_0154_s1` ausgewählt. Läuft auch nicht besser.

**Reply by caco3** [reply]
hmm, ev. hat @haverland noch einen tip, wieso der oberste analoge zeiger manchmal zu hoch ist.

**Reply by fm-maniac** [reply]
Es ist nicht nur der oberste analoge Zeiger. Mesistens ist es die zweite und dritte Nachkommastelle.

**Reply by fm-maniac** [reply]
Wie man hier sieht läuft die Erkennung durchaus perfekt.

![2023-02-20 17_31_48-](https://user-images.githubusercontent.com/113193480/220160065-fae78c78-aa64-40ab-9e75-993fdca0319c.png)

**Reply by caco3** [reply]
Gemäss https://jomjol.github.io/neural-network-analog-needle-readout/ sollte dein Zeigertyp eigentlich gut trainiert sein.

**Reply by ghost** [reply]
Das liegt leider am Gerät selbst. Schaut euch mal die Zeiger an:
<img width="240" alt="image" src="https://user-images.githubusercontent.com/412645/220170637-34d8d85c-c88d-4dc0-be76-2ecc32123113.png">

Wenn der vorhergehende Zeiger auf "5" ist, sollte der nachfolgende Zeiger zwischen 2 Ziffern sein. Das ist hier total nicht. Deshalb ist der Postprozess auch so falsch. Ich habe keine Idee wie man so etwas korrigieren sollte. Nur über "neg. Rate" und "MaxRateValue" die fehlerhaften erkennen uns ausgrenzen.

**Reply by fm-maniac** [reply]
Das ist merkwürdig. Hier haben die Zeiger eine gute Korrelation. 
![Screenshot_20230220_191637_org mozilla firefox_edit_11480433003455](https://user-images.githubusercontent.com/113193480/220177186-c05bb2d3-8df9-4345-95a0-278100b0e0b4.jpg)
Evtl. hängt einer der Zeiger sporadisch. Ich habe das über "neg. Rate" und "MaxRateVakue" ausgegrenzt. Nur resultiert das darin, dass ich lange Zeit keine Daten bekomme.

**Reply by ghost** [reply]
In diesem Fall ist es wirklich korrekt. Da wir aber im letzten halben Jahr nur diese Anfrage hatten, was innerhalb Analog betrifft, würde ich da jetzt nichts weiter unternehmen. 
Eigentlich hatten wir nur noch Probleme beim Analog-Digit Übergang.

PS: Auch wenn ich damit penetrant bin. Die Digital-Modelle können den "Fenterrand" der Zahl nicht gut erkennen. Daher besser den Rand oben und unten reduzieren (lock Aspekt ratio ausschalten)

**Reply by fm-maniac** [reply]
Kurz zu den Digits. Die wurden bisher gut erkannt. Hier hatte ich schon den Nulldurchgang von 309 auf 310m3 ohne Probleme. Die ROIs sind wie nachstehend eingerichtet. Der innere Rahmen umfasst eigentlich nur die Ziffer. Sollte ich den noch enger ziehen? 
![Screenshot_20230220_201359](https://user-images.githubusercontent.com/113193480/220184335-dc9ae99a-9599-4949-a845-eac1e7005eab.jpg)
Zurück zu den analogen Anzeigen. Es scheint als eile die x0.01 irgendwie nach. Müsste deutlich weiter an der 5 sein. Auch auf obigem Foto zu sehen. Keine Chance das in den Griff zu bekommen?

**Reply by caco3** [reply]
Wenn es ein systematische Fehler ist (sprich, ein Zeiger immer gleichviel Grad verdreht ist), könnte man das durch eine Erweiterung der Software korrigieren. Du könntest dir eine eigene Software dafür anpassen. Wenn man das generisch implementieren müsste, wäre das einiges an Aufwand.

**Reply by ghost** [reply]
Das Problem ist, dass es für User die Konfiguration immer komplexer wird.

@fm-maniac Persönlich würde ich 0.0001-0.01 nicht auslesen aber "extended Resolution" einstellen. Dann wäre der Zähler auf 10l genau.

**Reply by fm-maniac** [reply]
Das wäre ein möglicher Workaround. Auch wenn ich mir eine etwas höhere Auflösung als 10l wünschen würde.

Was ich aber nicht verstehe ist, dass wie zuvor beschrieben die Recognition korrekt auf 2.7 erkennt, der Postprozess aber fehlerhaft auf 3. Das beobachte ich regelmäßig, mit den fehlerhaft zu hohen Readings.

**Reply by ghost** [reply]
Es gibt eine Sonderbehandlung +/- 2 bei Analog bei Nulldurchgang, um Ungenauigkeiten auszugleichen. Oder bist du bei digital? Dort ist genau 0.2 eine Sonderbehandlung. Wenn du also dort 2.7 hast, kommst du aus dem Postprozess raus.
Da hilft nur https://jomjol.github.io/AI-on-the-edge-device-docs/ROI-Configuration/.

**Reply by ghost** [reply]
![](https://user-images.githubusercontent.com/412645/199028748-c48ef5bb-a8d4-4c77-9faf-763e6cf77351.png)

**Reply by fm-maniac** [reply]
Ich war schon bei den analogen Zeigern. Die Digits betrffend ziehe ich die ROIs nochmal enger um die Ziffer.
Das mit der Sonderbehandlung bzgl. der analogen Zeiger habe ich nicht verstanden. Sorry!

**Comment by fm-maniac** [comment]
Ich habe es noch nicht ganz aufgegeben, da ich nicht glaube das es an einem hängenden Zeiger liegt. Denn die Fehler treten an unterschiedlichen Nachkommastellen auf. Siehe CSV anbei. Da werden komplett falsche Werte erkannt. Auch bekomme ich in der recognition nach wie vor korrekte Werte, in der Auswertung aber nicht.

![2023-02-22 16_24_43-Window](https://user-images.githubusercontent.com/113193480/220676530-bc362ccf-bf43-4279-883d-09005789df1d.png)
![2023-02-22 16_25_25-Window](https://user-images.githubusercontent.com/113193480/220676536-8f8611a3-4205-415d-a59f-a0f8c674bdad.png)

[data_2023-02-22.csv](https://github.com/jomjol/AI-on-the-edge-device/files/10805666/data_2023-02-22.csv)

**Reply by ghost** [reply]
9.7 => wir befinden uns in Transition (Nulldurchlauf)
3.2 => wir runden auf 3 (2.8...3.2) wg. der Ungenauigkeit der neuronalen Netze
Jetzt wird nachgeschaut und festgestellt, dass der Nulldurchlauf noch nicht stattgefunden hat (9.7)
also keine 3.2 (bzw. 3) sondern noch eine 2


Und vermutlich ist es sogar eine 2.9, oder?
